srml/rolling-window/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Rolling Window module
+//!
+//! ## Overview
+//!
+//! The Rolling Window Module is similar to `simple moving average` except
+//! that it just reports the number of occurrences in the window instead of
+//! calculating the average.
+//!
+//! It is mainly implemented to keep track of misbehaviors and only the take
+//! the last `sessions` of misbehaviors into account.
+//!
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(test)]
+mod mock;
+
+use srml_support::{
+	StorageMap, EnumerableStorageMap, decl_module, decl_storage,
+	traits::WindowLength
+};
+use parity_codec::Codec;
+use sr_primitives::traits::MaybeSerializeDebug;
+use srml_session::SessionIndex;
+
+/// Rolling Window trait
+pub trait Trait: system::Trait {
+	/// Kind to report with window length
+	type Kind: Copy + Clone + Codec + MaybeSerializeDebug + WindowLength<u32>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as RollingWindow {
+		/// Misbehavior reports
+		///
+		/// It stores every unique misbehavior of a kind
+		// TODO [#3149]: optimize how to shrink the window when sessions expire
+		MisconductReports get(kind): linked_map T::Kind => Vec<SessionIndex>;
```

Comment:
> Why not mapping kind => hash instead? I think we are losing information that may be useful.

Because we are using `SessionIndex` to remove/filter out old misbehaviors and all hashes are stored in `BondingUniqueness` so we shouldn't be losing information. Ok?

Reply:
I was thinking in case we want to get the misbehaviors (hashes) of a kind, looks not possible. For filtering performance could be `Kind -> (SessionIndex, Hash)`. But yeah, I don't know if it's worth.

Reply:
Right I see, it more or less defeats the purpose of the rolling window but it might be useful. Let's keep it out of this PR and revisit again if we would need it, ok?

Reply:
Sure.